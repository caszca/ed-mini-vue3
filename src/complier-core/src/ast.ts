export const enum NodeType {
  INTERPOLATION,
  SIMPLE_EXPRESSION,
  ELEMENT,
  TEXT,
  ROOT, //transform中使其根节点也能遍历其children
}
