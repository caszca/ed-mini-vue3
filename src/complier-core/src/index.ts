//总导出出口
export * from "./compile";
