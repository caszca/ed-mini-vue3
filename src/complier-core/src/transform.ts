export function transform(ast, options = {}) {
  const context = createTransformContext(ast, options);
  traverseNode(ast, context);
}

function createTransformContext(ast: any, options: any) {
  return {
    ast,
    nodeTransforms: options.nodeTransforms || [],
  };
}

//深度优先遍历ast(抽象语法树)
function traverseNode(node, context: any) {
  const { nodeTransforms } = context;
  for (const plugin of nodeTransforms) {
    plugin(node);
  }
  const { children } = node;
  if (children) {
    for (const childNode of children) {
      traverseNode(childNode, context);
    }
  }
}
