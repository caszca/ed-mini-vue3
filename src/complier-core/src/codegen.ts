//负责将ast(抽象语法树)转换为render函数

import { NodeType } from "./ast";

function handle(val: any) {
  return `${val} : _${val}`;
}

function createCodegenContext(ast) {
  const context = {
    code: "",
    push(source) {
      context.code += source;
    },
    ast,
  };
  return context;
}

export function generate(ast) {
  console.log("-----", ast.children[0]);
  const context = createCodegenContext(ast);
  const { push } = context;
  handleImport(ast, push);
  const functionName = "render";
  const args = ["_ctx", "_cache"];
  push(`return function ${functionName}(${args.join(", ")}){ return `);
  codegen(ast.children[0], context);
  push(" }");
  return context;
}

//处理import情况
function handleImport(ast: any, push) {
  const { parameter } = ast;
  if (parameter.size) {
    let par: string[] = [];
    for (const val of parameter) {
      par.push(handle(val));
    }
    push(`const { ${par.join(", ")} } = vue`);
    push("\n");
  }
}

function codegen(rootNode, context) {
  switch (rootNode.type) {
    case NodeType.TEXT:
      genText(rootNode, context);
      break;
    case NodeType.INTERPOLATION:
      genInterpolation(rootNode, context);
      break;

    case NodeType.SIMPLE_EXPRESSION:
      genSimpleExpression(rootNode, context);
      break;

    case NodeType.ELEMENT:
      genElement(rootNode, context);
      break;

    case NodeType.COMPOUND_EXPRESSION:
      genCompound(rootNode, context);
      break;
    default:
      break;
  }
}

function codegenChildren(nodes, context) {
  if (nodes.length) {
    for (const node of nodes) {
      codegen(node, context);
    }
  }
}

function genText(node: any, context) {
  const { push } = context;
  push(`'${node.content}'`);
}

function genSimpleExpression(node: any, context) {
  const { push } = context;
  push(`${node.content}`);
}

function genInterpolation(node: any, context) {
  const { push } = context;
  push("_toDisplayString(");
  codegen(node.content, context);
  push(")");
}

function genElement(node, context) {
  const { push } = context;
  push(`_createElementVNode(`);
  genNodeList(node, context);
  push(")");
}

function genNodeList(node, context) {
  const { push } = context;
  const { tag, props, children } = node;
  push(`'${tag}', ${props || null}, `);
  console.log(children, "----------");
  children ? codegenChildren(children, context) : push("null");
}

function genCompound(node, context) {
  codegenChildren(node.children, context);
}
//return function render(_ctx, _cache, $props, $setup, $data, $options) { return "hi" }
