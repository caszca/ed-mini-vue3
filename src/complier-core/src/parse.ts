import { NodeType } from "./ast";

const enum TagType {
  Start,
  End,
}

export function baseParse(content) {
  const context = createParserContext(content);
  return createRoot(parseChildren(context, ""));
}

//解析children
function parseChildren(context, parentTag) {
  const nodes: any = [];
  let node;
  while (!isEnd(context, parentTag)) {
    const { source } = context;
    //解析插值
    if (source.startsWith("{{")) {
      node = parseInterpolation(context);
    } else if (source[0] == "<" && /[a-z]/i.test(source[1])) {
      //解析HTML元素
      node = parseElement(context);
    } else {
      //处理text
      node = parseText(context);
    }
    nodes.push(node);
  }
  return nodes;
}

//判断是否结束
function isEnd(context, parentTag) {
  const { source } = context;
  return !source || source.startsWith(`</${parentTag}>`);
}

//解析普通文本
function parseText(context) {
  let { source } = context;
  const delimiter = ["{{", "<", "</"];
  let endIndex = source.length;
  for (const key of delimiter) {
    let index = source.indexOf(key);
    if (index >= 0 && index < endIndex) {
      endIndex = index;
    }
  }

  if (endIndex == -1) {
    source = parseTextSlice(context, source.length);
  } else {
    source = parseTextSlice(context, endIndex);
  }

  return {
    type: NodeType.TEXT,
    content: source,
  };
}

function parseTextSlice(context, index) {
  let { source } = context;
  context.source = source.slice(index);
  source = source.slice(0, index);
  return source;
}

//解析HTML元素，<div></div>——>{tag:"div",type:元素}
function parseElement(context) {
  const { source } = context;
  //处理开始标签
  const result: any = parseTag(context, TagType.Start);
  result.children = parseChildren(context, result.tag);
  //处理结尾标签
  parseTag(context, TagType.End);
  return result;
}

function parseTag(context, type) {
  const { source } = context;
  const match: any = /^<\/?([a-z]*)/i.exec(source);
  const tag = match[1];
  context.source = source.slice(match[0].length + 1);
  if (type == TagType.End) return;
  return {
    type: NodeType.ELEMENT,
    tag,
  };
}

//解析插槽功能,将{{ message }}——>message,注意返回结果结构
function parseInterpolation(context) {
  const { source } = context;
  const openDelimiter = "{{";
  const closeDelimiter = "}}";
  const start = source.indexOf(openDelimiter);
  const end = source.indexOf(closeDelimiter);
  const content = source.slice(start + openDelimiter.length, end).trim();
  context.source = source.slice(end + closeDelimiter.length); //parse过的字符串就截掉，剩下后面的继续parse。

  return {
    type: NodeType.INTERPOLATION,
    content: {
      type: NodeType.SIMPLE_EXPRESSION,
      content,
    },
  };
}

function createRoot(children) {
  return {
    children,
  };
}

function createParserContext(content) {
  return {
    source: content,
  };
}
