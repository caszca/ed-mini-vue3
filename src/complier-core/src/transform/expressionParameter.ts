//transform插件，用于处理遇到插槽节点时，所需添加参数等

import { NodeType } from "../ast";
export function transformExpression(node) {
  if (node.type == NodeType.INTERPOLATION) {
    node.content.content = "_ctx." + node.content.content;
  }
}
