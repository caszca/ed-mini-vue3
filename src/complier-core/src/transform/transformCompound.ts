//处理相近的text与插槽
//不存在相近的节点是同一种类型，比如都是text与插槽
import { NodeType } from "../ast";

const addConst = {
  type: NodeType.SIMPLE_EXPRESSION,
  content: " + ",
};

export function transformCompound(node, context) {
  if (node.type == NodeType.ELEMENT) {
    const { children } = node;
    for (let i = 0; i < children.length; i++) {
      const node = children[i];
      if (isText_Interpolation(node)) {
        let container: any = null;
        for (let j = i + 1; j < children.length; j++) {
          const nodej = children[j];
          if (isText_Interpolation(nodej)) {
            //相近节点为text与插槽（不区分先后）
            if (!container) {
              //初次F
              container = children[i] = {
                type: NodeType.COMPOUND_EXPRESSION,
                children: [node],
              };
            }
            console.log("-----", j);
            children.splice(j, 1); //删掉已经添加进去的节点
            j--; //splice引起的数组后续往前移
            container.children.push(addConst);
            container.children.push(nodej);
          } else {
            //遇到非上述节点退出内层循环，继续外层循环
            container = null;
            break;
          }
        }
      }
    }
  }
}
//判断节点是否为插槽或文本text
function isText_Interpolation(node) {
  return node.type == NodeType.INTERPOLATION || node.type == NodeType.TEXT;
}
