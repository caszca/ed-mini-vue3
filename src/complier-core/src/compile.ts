import { generate } from "./codegen";
import { baseParse } from "./parse";
import { transform } from "./transform";
import { transformExpression } from "./transform/expressionParameter";
import { transformCompound } from "./transform/transformCompound";

//将参数template转化为一个render字符串
export function baseCompile(template) {
  const ast: any = baseParse(template);
  transform(ast, {
    nodeTransforms: [transformExpression, transformCompound],
  });

  return generate(ast);
}
