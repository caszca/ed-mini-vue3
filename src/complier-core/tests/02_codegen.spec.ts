import { generate } from "../src/codegen";
import { baseParse } from "../src/parse";
import { transform } from "../src/transform";
import { transformExpression } from "../src/transform/expressionParameter";
import { transformCompound } from "../src/transform/transformCompound";
describe("codegen", () => {
  it("string", () => {
    const ast = baseParse("hi");
    transform(ast);
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });

  it("interpolation", () => {
    const ast = baseParse("{{message}}");
    transform(ast, {
      nodeTransforms: [transformExpression],
    });
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });

  it.only("element", () => {
    const ast: any = baseParse("<div>hi,{{message}}</div>");
    transform(ast, {
      nodeTransforms: [transformExpression, transformCompound],
    });

    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });
});
