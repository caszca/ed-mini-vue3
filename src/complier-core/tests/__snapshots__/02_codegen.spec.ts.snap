// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element 1`] = `
"const { createElementVNode : _createElementVNode, toDisplayString : _toDisplayString } = vue
return function render(_ctx, _cache){ return _createElementVNode('div', null, 'hi,' + _toDisplayString(_ctx.message)) }"
`;
